### PR TITLE
Fix for shared axes diverging after setting tick markers

### PR DESCRIPTION
**Pull request by shaynair**
## PR Summary

This fixes issue #8946. 
A previous pull request #10651 was made, but I erroneously rebased to master which messed up the commit history there.

Example plot:

Associated code snippet to reproduce the above plot:
```python
import matplotlib.pyplot as plt
fig, (ax1, ax2) = plt.subplots(2, sharex=True)
ax1.plot(range(0, 10), range(0, 10))
ax2.plot(range(0, 10), range(0, 10))
plt.xticks(range(0, 30))
plt.show()
```

The two plots should be identical as they share the x-axis, but clearly aren't. This is a fix for this.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by shaynair** [comment]
Thanks for the feedback @efiring, just a few concerns:

- Would the `auto` parameter also be delegated to `set_view_interval` since it also needs to be propagated to the shared axes?
- I'm having trouble making `Axis3D` compatible with the `set_view_interval` change. `Axis3D` uses `v_interval` whereas `Axes3D.set_xlim` uses `xy_viewLim.intervalx` which are different. How would I add the `emit` parameter to `Axis3D.set_view_interval` without it breaking previous test cases? And how would it be called from `Axes3D.set_xlim`?

**Comment by shaynair** [comment]
Alright, I've incorporated the feedback given @efiring. Please let me know what you think.

File: lib/matplotlib/axis.py [hunk 1]
```diff
@@ -2059,18 +2060,24 @@ def get_view_interval(self):
...
 
-    def set_view_interval(self, vmin, vmax, ignore=False):
+    def set_view_interval(self, vmin, vmax, ignore=False, emit=True,
+                          auto=False):
         """
         If *ignore* is *False*, the order of vmin, vmax
         does not matter; the original axis orientation will
         be preserved. In addition, the view limits can be
-        expanded, but will not be reduced.  This method is
+        expanded, but will not be reduced.
```
Comment (jklymak) [inline lib/matplotlib/axis.py]: OK, while we are here, can we make these docstrings numpydoc compliant?  

This file has almost no compliant docstrings (apparently it hasn't attracted the attention of @timhoffm, *or* he already has a PR in that hasn't been merged.)  

As a random example of whats needed have a look at: https://github.com/matplotlib/matplotlib/blob/b6acd0402e6120693c0ff2f5cd345c208bde29f0/lib/matplotlib/figure.py#L1060

If you don't feel like doing this, its probably fine, but it'd help..

Comment (shaynair) [reply lib/matplotlib/axis.py]: Should I make the entire file have compliant docstrings, or just the methods that I changed?

Comment (jklymak) [reply lib/matplotlib/axis.py]: ha ha - go for the whole file if you want, but thats *totally* unfair.  Just the strings you changed :wink:

Comment (jklymak) [reply lib/matplotlib/axis.py]: plus, if you think thats too annoying, don't feel obliged.  Someone will hit the whole file soon I'm sure...

Comment (timhoffm) [reply lib/matplotlib/axis.py]: Did someone call me? :wink:

@shaynair No seriously, only update the docstrings as part of this PR if you are touching them anyway. It simplifies the reviews if they are not too large and cover only one aspect. Also, keeping bugfixes and major docstring cleanups separate makes possible backports easier. That said, you are very welcome to open a separate PR for further docstring changes in the file.

**Comment by jklymak** [comment]
This is still a bug in master, but will require some rebasing.  Punting to 3.2, but @shaynair keep pushing this if you want it to go in!

**Comment by jklymak** [comment]
@shaynair sorry this has taken a while to get approval.  Unfortunately the way our approval system works, stuff needs to be re-pinged frequently to attract attention.  

This needs a rebase, and then we can ask @efiring to look again.

**Comment by timhoffm** [comment]
Tried to rebase this, but I'm afraid this is beyond repair. Too much changed in the meantime, so that it would be easier to rewrite from scratch. The test could be recovered to document the problem and be a starting point for a fresh solution.

@shaynair I'm very sorry, this got lost and we couldn't bring it into matplotlib.

**Comment by timhoffm** [comment]
Test extracted, adapted and posted into the original issue: https://github.com/matplotlib/matplotlib/issues/8946#issuecomment-522268604

**Pull request closed without merging**